src/auth/cephx/CephxSessionHandler.cc

```diff
@@ -181,3 +181,47 @@ int CephxSessionHandler::check_message_signature(Message *m)
   return 0;
 }
 
+int CephxSessionHandler::sign_bufferlist(bufferlist &in, bufferlist &out)
+{
+  char exp_buf[CryptoKey::get_max_outbuf_size(in.length())];
+
+  try {
+    const CryptoKey::in_slice_t sin{in.length(),
+        reinterpret_cast<const unsigned char *>(in.c_str())};
+    const CryptoKey::out_slice_t sout{
+        sizeof(exp_buf),
+        reinterpret_cast<unsigned char *>(&exp_buf)};
+    key.encrypt(cct, sin, sout);
+  }
+  catch (std::exception &e) {
+    lderr(cct) << __func__ << " failed to encrypt signature block" << dendl;
+    return -1;
+  }
+
+
+  out.append(exp_buf, sizeof(exp_buf));
+
+  return 0;
+}
+
+int CephxSessionHandler::encrypt_bufferlist(bufferlist &in, bufferlist &out) {
+  std::string error;
+  try {
+    connection_secret.encrypt(cct, in, out, &error);
```

Comment:
For AES-CBC this starts with new IV each time, so each message in a session is encrypted independently from other messages. I see two drawbacks:
* IIUC we need to take care about e.g. replay attacks on our own,
* hurting performance especially in conjunction with no guarantees for thread-safety (PK11/EVP context needs to created and destroyed each time). Rectifying this issue would open a few new possibilities.

Do we have the guarantee that messages in a session are processed in-order?

Reply:
yeah, it's all ordered, so we should preserve teh IV (or wahtever the term is)